lib/ron.nix

```diff
@@ -0,0 +1,284 @@
+    toJSON
+  ;
+
+  RON = let
+    tuple = elements: {
+      __RONType = "tuple";
+      inherit elements;
+    };
+    tupleWithType = type: elements: {
+      __RONTupleType = type;
+    } // (tuple elements);
+
+    struct = attrSet: attrSet;
+    structWithType = type: attrSet: {
+      __RONStructType = type;
+    } // (struct attrSet);
+
+    list = list: list;
+
+    map = mappings: {
+      __RONType = "map";
+      inherit mappings;
+    };
+
+    identifier = name: {
+      __RONType = "identifier";
+      inherit name;
+    };
+
+    some = object: { __RONType = "some"; value = object; };
+    none = { __RONType = "none"; };
+
+    unit = null;
+
+    typeOf = object:
+      if object == null then "unit"
+      else if isList object then "list"
+      else if isAttrs object then object.__RONType or "struct"
+      else if isBool object then "boolean"
+      else if isString object then "string"
+      else if isInt object then "int"
+      else throw "Failed to get RON type of ${toJSON object}";
+
+    types = {
+      # option types.  RON structs and lists are isomorphic to Nix
+      # attrSets and lists, but optionals and maps are unfortunately
+      # special
+
+      optional = of: lib.mkOptionType {
+        name = "RONOptional";
+        description = "Either lib.RON.none or lib.RON.Some(${of.name})";
+        descriptionClass = "conjunction";
+        check = value:
+          let
+            rt = typeOf value;
+          in
+            rt == "none" || (rt == "some" && of.check value.value);
+        merge = lib.mergeEqualOption;
+        emptyValue = none;
+      };
+
+      map = to: lib.mkOptionType {
+        name = "RONMap";
+        description = "lib.RON.map from anything to ${to.name}";
+        descriptionClass = "composite";
+        check = value: typeOf value == "map";
+        merge = loc: defs:
+          let
+            doit = list:
+              if length list == 1 then head list
+              else let
+                x = (head list).mappings;
+                y = (doit (tail list)).mappings;
+              in
+                map ((builtins.map (xe: let
+                  ye = lib.findFirst (ye: xe.key == ye.key) null y;
+                in
+                  if ye == null then xe else { key = xe.key; value = to.merge loc [xe ye]; })
+                  x)
+                ++ (filter (ye: (lib.findFirst (xe: xe.key == ye.key) null x) == null) y));
+          in
+            doit (lib.getValues defs);
+        emptyValue = map [];
+      };
+    };
+  in {
+    inherit tuple tupleWithType struct structWithType;
+    inherit list map identifier some none unit typeOf;
+    inherit types;
+  };
+
+  fromRON = string: let
+    len = stringLength string;
+
+    assemble = extract: init: plus: rs:
+      let
+        doit = rs:
+          if length rs == 0 then init
+          else (plus (extract (head rs)) (doit (tail rs)));
+      in
+        doit rs;
+    assembleString = assemble (_:_) "" (x: y: x + y);
```

Comment:
you can use lib.id and builtins.add here (of course this is just a small nitpick that doesn't affect much)